modules/secure-logging/slog.c

```diff
@@ -1991,3 +1917,274 @@ gboolean validFileNameArg(const gchar *option_name, const gchar *value, gpointer
+                evt_tag_long("Code ", SLOG_FILE_GENERAL_ERROR),
+                evt_tag_str("Reason ", "General file stream error")
+               );
+    }
+}
+
+// Error handler for memory errors
+void handleGPtrArrayMemoryError(const char *file, int line, int code, void *object)
+{
+  GPtrArray *pa =(GPtrArray *)object;
+  const gchar *msg1 = "Premature memory release of GPtrArray";
+  const gchar *msg2 = "GPtrArray address is NULL";
+  const gchar *msg;
+
+  if(pa != NULL)
+    {
+      g_ptr_array_free(pa, TRUE);
+      msg = msg1;
+    }
+  else
+    {
+      msg = msg2;
+    }
+  msg_error(SLOG_ERROR_PREFIX,
+            evt_tag_str("File: ", file),
+            evt_tag_long("Line: ", line),
+            evt_tag_long("Code: ", code),
+            evt_tag_str("Reason: ", msg)
+           );
+}
+
+// Retrieve counter from encrypted log entry
+gboolean getCounter(GString *entry, guint64 *logEntryOnDisk)
+{
+  // Interpret the first COUNTER_LENGTH+1 characters
+  char ctrbuf[COUNTER_LENGTH+1];
+  memcpy(ctrbuf, entry->str, COUNTER_LENGTH);
+  ctrbuf[COUNTER_LENGTH] = 0;
+
+  gsize outLen;
+  guchar *ctr = g_base64_decode((const char *)ctrbuf, &outLen);
+  gboolean ret;
+
+  if (outLen != sizeof(guint64))
+    {
+      msg_error("[SLOG] ERROR: Cannot derive integer value from counter field",
+                evt_tag_str("Log entry number", (const char *)ctr));
+      ret = FALSE;
+    }
+  else
+    {
+      memcpy(logEntryOnDisk, ctr, sizeof(guint64));
+      ret = TRUE;
+    }
+
+  g_free(ctr);
+  return ret;
+}
+
+// Check whether value is contained in table
+gboolean tableContainsKey(GHashTable *table, guint64 key)
+{
+  if(table == NULL)
+    {
+      msg_error(SLOG_ERROR_PREFIX,
+                evt_tag_str("File: ", __FILE__),
+                evt_tag_long("Line: ", __LINE__),
+                evt_tag_long("Code: ", SLOG_MEM_ALLOC_ERROR),
+                evt_tag_str("Reason: ", "Table == NULL")
+               );
+      return FALSE;
+    }
+
+  // Convert to string
+  char keystr[CTR_LEN_SIMPLE+1];
+  snprintf(keystr, CTR_LEN_SIMPLE+1, "%"G_GUINT64_FORMAT, key);
+
+  return g_hash_table_contains(table, keystr);
+}
+
+// Add new value to table
+gboolean addValueToTable(GHashTable *table, guint64 value)
+{
+  if(table == NULL)
+    {
+      msg_error(SLOG_ERROR_PREFIX,
+                evt_tag_str("File: ", __FILE__),
+                evt_tag_long("Line: ", __LINE__),
+                evt_tag_long("Code: ", SLOG_MEM_ALLOC_ERROR),
+                evt_tag_str("Reason: ", "Table == NULL")
+               );
+      return FALSE;
+    }
+
+  // Create new key to string and use value as key
+  char *key = g_new0(char, CTR_LEN_SIMPLE+1);
+  snprintf(key, CTR_LEN_SIMPLE+1, "%"G_GUINT64_FORMAT, value);
+
+  return g_hash_table_insert(table, key, (gpointer)value);
+}
+
+// Get a single line from a log file
+GString *getLogEntry(SLogFile *f)
+{
+  GString *line = g_string_new(NULL);
+
+  if(!read_line_from_file(f, line))
```

Comment:
`read_line_from_file` could return with a GString instead of a gboolean, that way, unnecessary allocations can be eliminated, like we have here